dali/base/dasess.cpp

```diff
@@ -2055,36 +2066,56 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
         passwordenc.clear();
         sessionToken = 0;
         signature.clear();
+        utcTimeStamp.clear();
+        userTimeStampSignature.clear();
     }
     void serialize(MemoryBuffer &mb)
     {
         mb.append(username).append(passwordenc);
     }
-    void serializeExtra(MemoryBuffer &mb)
+    void serializeSignature(MemoryBuffer &mb)
     {
         if (queryDaliServerVersion().compare("3.14") >= 0)
         {
-            if (signature.isEmpty())
+            if (pDSM == nullptr)
+                pDSM = createDigitalSignatureManagerInstanceFromEnv();
+
+            //Serialize timestamp and signature of "username;timestamp"
+            //Dali will use this to ensure request came from
+            //valid authenticated user within a reasonable timeframe
+            if (pDSM && pDSM->isDigiSignerConfigured())
             {
-                if (pDSM == nullptr)
-                    pDSM = createDigitalSignatureManagerInstanceFromEnv();
-                if (pDSM && pDSM->isDigiSignerConfigured())
-                    pDSM->digiSign(username, signature);//Set user's digital signature
+                CDateTime now;
+                now.setNow();
+                StringBuffer timeStr;
+                now.getString(timeStr, false);//get UTC timestamp
+                VStringBuffer toSign("%s;%s", username.str(), timeStr.str());
+
+                StringBuffer userTimeStampSignature;
+                pDSM->digiSign(toSign, userTimeStampSignature);//Sign "username;timeStamp"
+
+                //Serialize the timestamped signature, and the timestamp object
+                mb.append(userTimeStampSignature.str());
+                now.serialize(mb);
             }
-            mb.append(signature);
         }
     }
     void deserialize(MemoryBuffer &mb)
     {
         mb.read(username).read(passwordenc);
     }
-    void deserializeExtra(MemoryBuffer &mb)
+    void deserializeSignature(MemoryBuffer &mb)
     {
-        signature.clear();
+        utcTimeStamp.clear();
+        userTimeStampSignature.clear();
         if (queryDaliServerVersion().compare("3.14") >= 0)
         {
             if (mb.remaining() > 0)
-                mb.read(signature);
+            {
+                mb.read(userTimeStampSignature);
+                if (mb.remaining() > 0)
```

Comment:
If the serialization format changes, the server version should normally change..
Otherwise you have varying different '3.14' serialization formats.